fvm/Grid/AnalyticBRadialGridGenerator.cpp

```diff
@@ -358,36 +358,342 @@ void AnalyticBRadialGridGenerator::InterpolateInputProfileToGrid(
+ * intervall is r = startingGuess +/- lengthScale, translated by a multiple of 2*lengthScale
+ * until the search interval contains a sign change
+ *
+ * See doc/notes/SPICoordinates.pdf for more information
  */
-real_t AnalyticBRadialGridGenerator::FindClosestApproach(
-    real_t /*x1*/, real_t /*y1*/, real_t /*z1*/,
-    real_t /*x2*/, real_t /*y2*/, real_t /*z2*/
+void AnalyticBRadialGridGenerator::GetRThetaPhiFromCartesian(real_t* r, real_t* theta, real_t* phi,
+    real_t x, real_t y, real_t z, real_t lengthScale, real_t startingGuessR
 ) {
-	throw FVMException("AnalyticBRadialGridGenerator: This module is currently incompatible with the SPI module.");
-    return 0;
+
+    real_t RMinusR0;
+    if(R0IsInf)
+        RMinusR0 = x;
+    else
+        RMinusR0 = hypot(R0+x, z)-R0;
+        
+	// Collect data for the given profiles of the shaping parameters
+	const real_t *kappaProf = this->providedProfiles->kappa;
+	const real_t *kappa_r = this->providedProfiles->kappa_r;
+	const len_t nkappa = this->providedProfiles->nkappa;
+	
+	const real_t *deltaProf = this->providedProfiles->delta;
+	const real_t *delta_r = this->providedProfiles->delta_r;
+	const len_t ndelta = this->providedProfiles->ndelta;
+	
+	const real_t *DeltaProf = this->providedProfiles->Delta;
+	const real_t *Delta_r = this->providedProfiles->Delta_r;
+	const len_t nDelta = this->providedProfiles->nDelta;
+	
+                
+    // We start by finding the flux surface label 'rmin' corresponding to cos(theta)=0 (using Newton iteration)
+    // and use it to determine in which quadrant we should look for the solution
+	real_t r_tmp=std::abs(y);
+	real_t Zdiff;// y in SPI coordinates is Z in cylindrical coordinates
+	real_t kappa;
+	real_t kappa_p;
+	
+	do{
+	    kappa = InterpolateInputProfileSingleExtrap(r_tmp, nkappa, kappaProf, kappa_r, 
+	            this->spline_kappa, gsl_acc_kappa);
+	    kappa_p = InterpolateInputProfileSingleDerivExtrap(r_tmp, nkappa, kappa_r, 
+	              this->spline_kappa, gsl_acc_kappa);
+	    Zdiff = std::abs(y)-r_tmp*kappa;
+	    r_tmp = r_tmp + Zdiff/(r_tmp*kappa_p+kappa);
+	    
+	    	
+	}while (std::abs(Zdiff)>lengthScale*CartesianCoordinateTol);
+	
+	real_t rmin=r_tmp;
+	
+	// Evaluate RMinusR0_crit corresponding to rmin
+	real_t delta;
+	real_t Delta; 
+    delta = InterpolateInputProfileSingleExtrap(rmin, ndelta, deltaProf, delta_r, 
+            this->spline_delta, gsl_acc_delta);
+    Delta = InterpolateInputProfileSingleExtrap(rmin, nDelta, DeltaProf, Delta_r, 
+            this->spline_Delta, gsl_acc_delta);
+    
+    real_t RMinusR0_crit=Delta-rmin*sin(delta);
+	
+	// Determine the quadrant
+	len_t quadrant;
+	if(RMinusR0>RMinusR0_crit && y>=0)
+	    quadrant=1;
+	else if(RMinusR0<RMinusR0_crit && y>=0)
+	    quadrant=2;
+    else if(RMinusR0<RMinusR0_crit && y<0)
+        quadrant=3;
+    else if(RMinusR0>RMinusR0_crit && y<0)
+        quadrant=4;
```

Comment:
You should configure your text editor to only use either tabs OR spaces. The current mixture of tabs and spaces (even on the same line) is bound to lead to confusion at some point.

Reply:
(in Python it's of course even worse, but I presume you have some way of ensuring consistency there?)